types/src/rsmt2d.rs

```diff
@@ -151,54 +144,144 @@ pub struct ExtendedDataSquare {
+                Namespace::from_raw(&share[..NS_SIZE])?
+            } else {
+                Namespace::PARITY_SHARE
+            };
+
+            if prev_ns.map_or(false, |prev_ns| ns < prev_ns) {
+                let axis_idx = match axis {
+                    AxisType::Row => row,
+                    AxisType::Col => col,
+                };
+                bail_validation!("Shares of {axis} {axis_idx} are not sorted by their namespace");
+            }
+
+            Ok(ns)
+        };
+
         // Validate that namespaces of each row are sorted
         for row in 0..eds.square_len() {
             let mut prev_ns = None;
 
             for col in 0..eds.square_len() {
-                let share = eds.share(row, col)?;
+                prev_ns = Some(check_share(row, col, prev_ns, AxisType::Row)?);
+            }
+        }
+        // Validate that namespaces of each column are sorted
+        for col in 0..eds.square_len() {
+            let mut prev_ns = None;
 
-                if share.len() != SHARE_SIZE {
-                    bail_validation!("share len ({}) != SHARE_SIZE ({})", share.len(), SHARE_SIZE);
-                }
+            for row in 0..eds.square_len() {
+                prev_ns = Some(check_share(row, col, prev_ns, AxisType::Col)?);
+            }
+        }
 
-                let ns = if is_ods_square(row, col, eds.square_len()) {
-                    Namespace::from_raw(&share[..NS_SIZE])?
-                } else {
-                    Namespace::PARITY_SHARE
-                };
+        Ok(eds)
+    }
 
-                if prev_ns.map_or(false, |prev_ns| ns < prev_ns) {
-                    bail_validation!("Shares of row {row} are not sorted by their namespace");
-                }
+    /// Create a new EDS out of the provided original data square shares.
+    ///
+    /// This method is similar to the [`ExtendedDataSquare::new`] but parity data
+    /// will be encoded automatically using the [`leopard_codec`]
+    ///
+    /// Shares should be provided in a row-major order.
+    ///
+    /// # Errors
+    ///
+    /// The same errors as in [`ExtendedDataSquare::new`] applies. The constrain
+    /// will be checked after the parity data is generated.
+    ///
+    /// Additionally, this function will propagate any error from encoding parity data.
+    pub fn from_ods(mut ods_shares: Vec<Vec<u8>>) -> Result<ExtendedDataSquare> {
+        let ods_width = f64::sqrt(ods_shares.len() as f64) as usize;
+        // this couldn't be detected later in `new()`
+        if ods_width * ods_width != ods_shares.len() {
+            return Err(Error::EdsInvalidDimentions);
+        }
 
-                prev_ns = Some(ns);
+        let eds_width = ods_width * 2;
+        let mut eds_shares = Vec::with_capacity(eds_width * eds_width);
+        // take rows of ods and interleave them with parity shares
+        for _ in 0..ods_width {
+            eds_shares.extend(ods_shares.drain(..ods_width));
```

Comment:
Do we need to check that shares from ODS actually have size of `SHARE_LEN`? Or `leopard_codec::encode` will take care of that?

Reply:
leopard will return an error then

Reply:
and even if not and encoding somehow worked, then `new` would

Reply:
https://github.com/eigerco/leopard-codec/blob/main/src/lib.rs#L42